rosbag2_py/src/rosbag2_py/_transport.cpp

```diff
@@ -140,50 +140,109 @@ class Player
     rclcpp::shutdown();
   }
 
+  static void cancel()
+  {
+    exit_ = true;
+    wait_for_exit_cv_.notify_all();
+  }
+
   void play(
     const rosbag2_storage::StorageOptions & storage_options,
     PlayOptions & play_options)
   {
-    auto reader = rosbag2_transport::ReaderWriterFactory::make_reader(storage_options);
-    auto player = std::make_shared<rosbag2_transport::Player>(
-      std::move(reader), storage_options, play_options);
-
-    rclcpp::executors::SingleThreadedExecutor exec;
-    exec.add_node(player);
-    auto spin_thread = std::thread(
-      [&exec]() {
-        exec.spin();
-      });
-    player->play();
-    player->wait_for_playback_to_finish();
-
-    exec.cancel();
-    spin_thread.join();
+    play_impl(storage_options, play_options, false);
   }
 
   void burst(
     const rosbag2_storage::StorageOptions & storage_options,
     PlayOptions & play_options,
     size_t num_messages)
   {
-    auto reader = rosbag2_transport::ReaderWriterFactory::make_reader(storage_options);
-    auto player = std::make_shared<rosbag2_transport::Player>(
-      std::move(reader), storage_options, play_options);
-
-    rclcpp::executors::SingleThreadedExecutor exec;
-    exec.add_node(player);
-    auto spin_thread = std::thread(
-      [&exec]() {
-        exec.spin();
+    play_impl(storage_options, play_options, true, num_messages);
+  }
+
+protected:
+  void play_impl(
+    const rosbag2_storage::StorageOptions & storage_options,
+    PlayOptions & play_options,
+    bool burst = false,
+    size_t burst_num_messages = 0)
+  {
+    auto old_sigterm_handler = std::signal(
+      SIGTERM, [](int /* signal */) {
+        rosbag2_py::Player::cancel();
+      });
+    auto old_sigint_handler = std::signal(
+      SIGINT, [](int /* signal */) {
+        rosbag2_py::Player::cancel();
       });
```

Comment:
I do not see those original handlers be called on anywhere, is that correct behavior? i was expecting that on signal SIGINT and SIGTERM, we can cancel the player and then original signal handlers should be called accordingly based on signal number? but as far as i can see, it reset the original handler during exit, never called?

Reply:
@fujitatomoya That is not easy to implement since I am not saving signal number and handling signals asynchronously.  
There is basically no information at the end of the Play or Burst method if the signal was triggered or not.

Reply:
@fujitatomoya Apparently, we do have a signal handler class on the rclcpp layer https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/signal_handler.hpp#L43-L55 but it is an inner class just for the rclcpp.  
I will try to see if I will be able to create something similar but lightweight for rosbag2 with low effort to be able to handle signals more systematically.

Reply:
@fujitatomoya I've added the call of the original signal handler for the `Player and `Recorder` classes in the https://github.com/ros2/rosbag2/pull/1557/commits/290933247fbc8a75140573c62a9edda73b1bb8a1. Please review.